### PR TITLE
implemented functions to control stepper motors for linear and rotary units

### PR DESCRIPTION
**Pull request by Ferran-RMS**

### CODE REVIEW

File: firmware/main.c [hunk 1]
```diff
@@ -709,11 +713,16 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
                                 start=false;
                                 chprintf(chp, "Pin 29 (Start) cleared! \r\n");
 			}
+			
                 break;
                 default: chprintf(chp, "Select one of the available modes \r\n");
 
 
         }
+//	palClearPad(pinPorts[pOpt1].gpio, pinPorts[pOpt1].pin); // Reset pins as its default state is high
```
Comment (ft) [inline firmware/main.c]: Personally, I really dislike  commented out code. Why's this commented out? If it's optional, it should be made like that, at compile time or at run time, whichever is preferred.

File: firmware/main.c [hunk 2]
```diff
@@ -834,6 +834,148 @@ static void cmd_motor_mode(BaseSequentialStream *chp, int argc, char *argv[]){
...
+/* Function for the Linear unit */
+
+static void cmd_motor_lu(BaseSequentialStream *chp, int argc, char *argv[]){
+        uint8_t i,pin2use,direction,start,gear_reduction=48,pOpt1, pOpt2; // Initialize the pins to be used //RL-D-50 has a gear reduction of 1:48
+        uint16_t num_pulses,u; // Unsigned short (0-65535)
+        char *dOpt = NULL, *pOpt = NULL, *mOpt = NULL; // *sOpt = NULL;
+        bool out1 = false, out2 = false; // Variables to store the state of the outputs
+
+        // Parsing
+        for(i = 0; i < argc; i++) {
```
Comment (ft) [inline firmware/main.c]: This looks like it should use getopt(3).

File: firmware/main.c [hunk 3]
```diff
@@ -834,6 +834,148 @@ static void cmd_motor_mode(BaseSequentialStream *chp, int argc, char *argv[]){
...
+                || ((strcmp(mOpt, "1") == 0) && (!pOpt || !dOpt))
+                || ((strcmp(mOpt, "2") == 0) && (!pOpt || !dOpt))
+                /*|| ((strcmp(mOpt, "3") == 0) && !sOpt)*/
+                /*|| ((strcmp(mOpt, "4") == 0) && !sOpt)*/
+|| !((strcmp(mOpt, "1") == 0) || (strcmp(mOpt, "2") == 0) || (strcmp(mOpt, "3") == 0) || (strcmp(mOpt, "4") == 0))
+        )goto exit_with_usage;
+
+        /* Pin definition for the pulse and direction outputs*/
+
+        for(i = 0; i < sizeof(pinPorts)/sizeof(pinPorts[0]); i++) {
```
Comment (ft) [inline firmware/main.c]: All these "if-in-loop" sections look very similar, which is usually a sign for missing abstractions. That should make these more readable too.

File: firmware/main.c [hunk 4]
```diff
@@ -615,7 +615,7 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
         palClearPad(pinPorts[pOpt2].gpio, pinPorts[pOpt2].pin);
 	palClearPad(pinPorts[start].gpio, pinPorts[start].pin);
 
-	num_pulses = atoi(pOpt) * gear_reduction; // Gets the right number of pulses to be done according to the gear reduction
+	num_pulses = atoi(pOpt) * gear_reduction * 2; // Gets the right number of pulses to be done according to the gear reduction, multiplied by 2
```
Comment (ft) [inline firmware/main.c]: There should be some description as to *why* this needs to be times-2. Also, style-nit, when lines get as long as this, the comment should go *before* the code and be made into a proper paragraph-style comment.

File: firmware/main.c [hunk 5]
```diff
@@ -633,6 +633,7 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
                         }
 			
 			chprintf(chp, "Pin to use: %d pinPorts size of: %d and number of pulses to produce %d \r\n",pin2use,i,num_pulses);
+			chThdSleepMilliseconds(200); // In order to have enough time to have a pulse for the start/enable signal
```
Comment (ft) [inline firmware/main.c]: Again, I'd put the comment in a paragraph-style comment before the code.

File: firmware/main.c [hunk 6]
```diff
@@ -553,17 +553,17 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
                         if(++i >= argc) continue;
                         mOpt = argv[i];  // Selects the operation mode
                 }
-		else if(strcmp(argv[i], "-s") == 0) {
```
Comment (ft) [inline firmware/main.c]: Again, in this commit, there is a bunch of commented out code. Like I wrote before, if it's obsolete, remove it; if it's optional make it optional.

File: firmware/main.c [hunk 7]
```diff
@@ -555,8 +555,13 @@ static void cmd_motor_rt(BaseSequentialStream *chp, int argc, char *argv[]){
...
         }
 
-        if (!dOpt || !pOpt || !mOpt)
-                goto exit_with_usage;
+        if (!mOpt
+		|| ((strcmp(mOpt, "1") == 0) && (!pOpt || !dOpt))
+		|| ((strcmp(mOpt, "2") == 0) && (!pOpt || !dOpt)) 
+		|| ((strcmp(mOpt, "3") == 0) && !sOpt)
+		|| ((strcmp(mOpt, "4") == 0) && !sOpt)
+|| !((strcmp(mOpt, "1") == 0) || (strcmp(mOpt, "2") == 0) || (strcmp(mOpt, "3") == 0) || (strcmp(mOpt, "4") == 0))
```
Comment (ft) [inline firmware/main.c]: The structure here looks a little funky. Long line, and closing paren right in front of the goto?

**Comment by ft** [comment]
Took a quick look and left a number of comments. I'm not quite familiar with the code base, so lots of my comments are style nits like "Why would you commit commented out code, when you have a revision control system to track changes?". There's also one instance, where I'm pretty sure you'd want to invest a little more time coming up with a proper abstraction in order to remove a sequence of loops, that look very copy-and-pasty.

File: firmware/main.c (hunk 2 above)
Comment (dpiegdon) [reply firmware/main.c]: there is no getopt in chibios, and i'd like to not use newlib. firmware already is rather large (50k bin) also the original codebase from chibios does the same, so this is just style mimicri.

**Comment by dpiegdon** [comment]
as a side note: i created the pull request with ferran to have the option to review the code. this was by no means meant to be commited right away. should have mentioned that.
I have started reviewing and fixing locally now.

**Comment by dpiegdon** [comment]
i finally started reviewing the code.
including my changes, these can be found in branch Ferran-RMS-cleanup.

**Comment by dpiegdon** [comment]
Ferran, can you take a look into the mentioned branch, Ferran-RMS-cleanup?
I have a question on cmd_motor_rotary() when using mode 1 or 2:
don't I need to set the pinEnable before, and clear it after executing the pulse train?
this was not done in the older version you wrote, but if I had to guess, I'd say we need this.
I marked the potential location with 'FIXME's.

**Comment by Ferran-RMS** [comment]
Hi David !
Is not needed, as in modes 1 and 2, is just waiting for a clk and direction signal in order to move.
The modes 3 and 4 require a enable signal only.

On the SMCi35, the so called "profiles" are just a description of what the motor controller should do on each of them. If a certain movement is programmed, then the motor controller only needs a trigger signal to execute the movement.

Just let me know:)

Cheers!

**Comment by dpiegdon** [comment]
ok!

**Pull request merged**